ipa-core/src/protocol/context/dzkp_malicious.rs

```diff
@@ -8,99 +8,108 @@ use ipa_step::{Step, StepNarrow};
+    helpers::{MpcMessage, MpcReceivingEnd, Role, SendingEnd, TotalRecords},
     protocol::{
         context::{
-            dzkp_validator::{DZKPBatch, Segment},
+            batcher::Batcher,
+            dzkp_validator::{Batch, Segment},
             prss::InstrumentedIndexedSharedRandomness,
-            Base, Context as ContextTrait, DZKPContext, InstrumentedSequentialSharedRandomness,
+            step::ZeroKnowledgeProofValidateStep,
+            Context as ContextTrait, DZKPContext, InstrumentedSequentialSharedRandomness,
+            MaliciousContext,
         },
-        prss::Endpoint as PrssEndpoint,
         Gate, RecordId,
     },
     seq_join::SeqJoin,
-    sync::Arc,
+    sync::{Arc, Mutex, Weak},
 };
 
+pub(super) type DzkpBatcher<'a> = Mutex<Batcher<'a, Batch>>;
+
 /// Represents protocol context in malicious setting when using zero-knowledge proofs,
 /// i.e. secure against one active adversary in 3 party MPC ring.
 #[derive(Clone)]
 pub struct DZKPUpgraded<'a> {
-    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
-    /// may operate with raw references and be more efficient
-    inner: Arc<DZKPUpgradedInner<'a>>,
-    gate: Gate,
-    total_records: TotalRecords,
+    batcher: Weak<DzkpBatcher<'a>>,
+    base_ctx: MaliciousContext<'a>,
 }
 
 impl<'a> DZKPUpgraded<'a> {
-    pub(super) fn new<S: Step + ?Sized>(
-        source: &Base<'a>,
-        malicious_step: &S,
-        batch: DZKPBatch,
-    ) -> Self
-    where
-        Gate: StepNarrow<S>,
-    {
+    pub(super) fn new(batch: &Arc<DzkpBatcher<'a>>, base_ctx: MaliciousContext<'a>) -> Self {
         Self {
-            inner: DZKPUpgradedInner::new(source, batch),
-            gate: source.gate().narrow(malicious_step),
-            total_records: source.total_records,
+            batcher: Arc::downgrade(batch),
+            base_ctx,
         }
     }
+
+    pub fn push(&self, record_id: RecordId, segment: Segment) {
+        self.with_batch(record_id, |batch| {
+            batch.push(self.base_ctx.gate().clone(), record_id, segment);
+        });
+    }
+
+    fn with_batch<C: FnOnce(&mut Batch) -> T, T>(&self, record_id: RecordId, action: C) -> T {
+        let batcher = self.batcher.upgrade().expect("Validator is active");
+
+        let mut batch = batcher.lock().unwrap();
+        let state = batch.get_batch(record_id);
+        (action)(&mut state.batch)
+    }
 }
 
 #[async_trait]
 impl<'a> DZKPContext for DZKPUpgraded<'a> {
-    fn is_verified(&self) -> Result<(), Error> {
-        if self.inner.batch.is_empty() {
-            Ok(())
-        } else {
-            Err(Error::ContextUnsafe(format!("{self:?}")))
-        }
-    }
-
-    fn push(&self, record_id: RecordId, segment: Segment) {
-        self.inner.batch.push(self.gate.clone(), record_id, segment);
+    async fn validate_record(&self, record_id: RecordId) -> Result<(), Error> {
```

Comment:
`validate_record` moves from the validator to the context. This matches what was done for the MAC implementation, and is desirable because we want the validator to be unclonable so that ownership follows a clear path all the way to the check that validation has happened when it is dropped.